util/buildinfo/buildinfo.go

```diff
@@ -79,7 +87,7 @@ func mergeSources(ctx context.Context, buildSources map[string]string, frontendS
 					break
 				}
 			}
-			if _, ok := mbs[sourceID.Reference.String()]; !ok {
+			if _, ok := mbs[sourceID.Reference.String()]; !ok && !hasFrontendSources {
```

Comment:
However here I check if frontend sources have been used before setting an llb source. This avoid having the same source with a ref digest being added.

I was also wondering if it would not have been better to set the user input (stage name) in the frontend though a llb description so we can link it with the llb source:

```go
func dfStage(name string) llb.ConstraintsOpt {
	return llb.WithDescription(map[string]string{
		"com.docker.dockerfile.v1.stage": name,
	})
}
```

Reply:
LLB sources should always be added. The only case is the alias set up from frontend. But I've not seen any indication that the current alias support is broken.

Reply:
It's currently broken when we added dependencies because LLB sources carry all sources (main and deps) so the solver while merging sources will also add deps sources not related to main and vice versa.

With this repro:

```Dockerfile
# base.Dockerfile
FROM alpine
ENV FOO=bar
ADD https://raw.githubusercontent.com/moby/moby/master/README.md /
RUN echo first > /out
```

```Dockerfile
# Dockerfile
FROM base AS build
RUN echo "foo is $FOO" > /foo
FROM busybox
COPY --from=build /foo /out /
```

We would have as result:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base": "input:base"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",
      "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
    },
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox@sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "http",
      "ref": "https://raw.githubusercontent.com/moby/moby/master/README.md",
      "pin": "sha256:419455202b0ef97e480d7f8199b26a721a417818bc0e2d106975f74323f25e6c"
    }
  ],
  "deps": {
    "base": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "docker.io/library/alpine:latest",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

`docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3` and `docker.io/library/busybox@sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578` should be removed but those refs don't match anything in deps. That's why at first I have normalized the ref with tag only in the dedup func to have a match, but it would break and remove refs digest that could be defined in frontend (dockerfile).

That's why I use this check on frontend sources existence so we would have:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base": "input:base"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "http",
      "ref": "https://raw.githubusercontent.com/moby/moby/master/README.md",
      "pin": "sha256:419455202b0ef97e480d7f8199b26a721a417818bc0e2d106975f74323f25e6c"
    }
  ],
  "deps": {
    "base": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "docker.io/library/alpine:latest",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

I agree that's not optimal, hence the comment https://github.com/moby/buildkit/pull/2684#discussion_r819932401 about maybe use llb description to link sources and context.

Reply:
I don't understand why the busybox is twice in the regular sources. But I think these cases can all be covered by the dedup function. Just the comparison can't be simple struct compare then. When there is ref already by tag in deps (or local sources although I don't understand where this case comes from) that has the same pin as ref by digest then the ref by digest can be ignored.

Reply:
Yes and also `https://raw.githubusercontent.com/moby/moby/master/README.md` should be in `deps.base`, not regular sources. I think our issue here is that LLB sources carry all sources and we don't have a way to know what source belongs to a regular or dep one.

Reply:
> I think our issue here is that LLB sources carry all sources and we don't have a way to know what source belongs to a regular or dep one.

Yes, and we have no way to know that. But we can just remove the source if it is already under a dep. Doesn't matter if it was used directly or only through a dep. For replay, we just need it under sources or under dep sources.

Reply:
> Yes, and we have no way to know that. ... For replay, we just need it under sources or under dep sources.

Ok then I don't see the point of having `deps` field in buildinfo struct? I think that's confusing and deps should be merged in buildinfo instead. WDYT?

```yaml
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base": "input:base",
    "build-arg:bar": "foo"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "http",
      "ref": "https://raw.githubusercontent.com/moby/moby/master/README.md",
      "pin": "sha256:419455202b0ef97e480d7f8199b26a721a417818bc0e2d106975f74323f25e6c"
    }
  ],
  "deps": {
    "base": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "docker.io/library/alpine:latest",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

would be

```yaml
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "build-arg:bar": "foo"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/alpine:latest",
      "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
    },
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "http",
      "ref": "https://raw.githubusercontent.com/moby/moby/master/README.md",
      "pin": "sha256:419455202b0ef97e480d7f8199b26a721a417818bc0e2d106975f74323f25e6c"
    }
  ]
}
```

Reply:
> I don't understand why the busybox is twice in the regular sources

For deps to be immutable we have to handle buildinfo in the [gateway](https://github.com/moby/buildkit/pull/2684/files#diff-b016f4bac6986ada033d4b3498671ff93059731d912904c5bfffb4c8e3358322R667-R709).

With the above example there is a first pass in the gateway for the `base` dep that will create its buildinfo:

```json
{
   "frontend": "dockerfile.v0",
   "sources": [
     {
       "type": "docker-image",
       "ref": "docker.io/library/alpine:latest",
       "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
     }
   ]
 }
```

As you can see we don't have `https://raw.githubusercontent.com/moby/moby/master/README.md` because LLB sources are always empty in gateway (`ref.BuildSource()`).

Then there is another pass in gateway for regular build too:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base": "input:base"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    }
  ],
  "deps": {
    "base": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "docker.io/library/alpine:latest",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

After that there is the final pass in the [solver](https://github.com/moby/buildkit/blob/0a7d8907ae7bc7262a8b61ea46741f7ec63c03ca/solver/llbsolver/solver.go#L164-L186). This one contains LLB sources:

```json
{
  "docker-image://docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",      
  "docker-image://docker.io/library/busybox@sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578",     
  "https://raw.githubusercontent.com/moby/moby/master/README.md": "sha256:419455202b0ef97e480d7f8199b26a721a417818bc0e2d106975f74323f25e6c"
}
```

And as we always add sources from LLB, they will be duplicated:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base": "input:base"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/alpine@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3",
      "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
    },
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox@sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "http",
      "ref": "https://raw.githubusercontent.com/moby/moby/master/README.md",
      "pin": "sha256:419455202b0ef97e480d7f8199b26a721a417818bc0e2d106975f74323f25e6c"
    }
  ],
  "deps": {
    "base": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "docker.io/library/alpine:latest",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

That's why [this condition](https://github.com/moby/buildkit/pull/2684/files#diff-8d246c4a422f0502d7589686322f61ad10370751b75f3a9471533b2a1f09b1d2R90) is in place so we don't add sources that have already been processed. Then we can accurately [dedup structs](https://github.com/moby/buildkit/pull/2684/files#diff-8d246c4a422f0502d7589686322f61ad10370751b75f3a9471533b2a1f09b1d2R217) without relying on pin/digest which could remove user pinned sources:

```json
{
  "frontend": "dockerfile.v0",
  "attrs": {
    "context:base": "input:base"
  },
  "sources": [
    {
      "type": "docker-image",
      "ref": "docker.io/library/busybox:latest",
      "pin": "sha256:b69959407d21e8a062e0416bf13405bb2b71ed7a84dde4158ebafacfa06f5578"
    },
    {
      "type": "http",
      "ref": "https://raw.githubusercontent.com/moby/moby/master/README.md",
      "pin": "sha256:419455202b0ef97e480d7f8199b26a721a417818bc0e2d106975f74323f25e6c"
    }
  ],
  "deps": {
    "base": {
      "frontend": "dockerfile.v0",
      "sources": [
        {
          "type": "docker-image",
          "ref": "docker.io/library/alpine:latest",
          "pin": "sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
        }
      ]
    }
  }
}
```

Maybe the gateway is not the right place to solve deps?